azuredevops/internal/service/permissions/resource_workitemquery_permissions.go

```diff
@@ -0,0 +1,192 @@
+package permissions
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/ahmetb/go-linq"
+	"github.com/google/uuid"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/workitemtracking"
+	"github.com/terraform-providers/terraform-provider-azuredevops/azuredevops/internal/client"
+	securityhelper "github.com/terraform-providers/terraform-provider-azuredevops/azuredevops/internal/service/permissions/utils"
+	"github.com/terraform-providers/terraform-provider-azuredevops/azuredevops/internal/utils/converter"
+)
+
+// ResourceWorkItemQueryPermissions schema and implementation for project permission resource
+func ResourceWorkItemQueryPermissions() *schema.Resource {
+	return &schema.Resource{
+		Create: ResourceWorkItemQueryPermissionsCreate,
+		Read:   ResourceWorkItemQueryPermissionsRead,
+		Update: ResourceWorkItemQueryPermissionsUpdate,
+		Delete: ResourceWorkItemQueryPermissionsDelete,
+		Schema: securityhelper.CreatePermissionResourceSchema(map[string]*schema.Schema{
+			"project_id": {
+				Type:         schema.TypeString,
+				ValidateFunc: validation.IsUUID,
+				Required:     true,
+				ForceNew:     true,
+			},
+			"path": {
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringIsNotWhiteSpace,
+				Optional:     true,
+				Required:     false,
+				ForceNew:     true,
+			},
+		}),
+	}
+}
+
+func ResourceWorkItemQueryPermissionsCreate(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+
+	sn, err := securityhelper.NewSecurityNamespace(clients.Ctx,
+		securityhelper.SecurityNamespaceIDValues.WorkItemQueryFolders,
+		clients.SecurityClient,
+		clients.IdentityClient)
+	if err != nil {
+		return err
+	}
+
+	aclToken, err := createWorkItemQueryToken(clients.Ctx, clients.WorkItemTrackingClient, d)
+	if err != nil {
+		return err
+	}
+
+	err = securityhelper.SetPrincipalPermissions(d, sn, aclToken, nil, false)
+	if err != nil {
+		return err
+	}
+
+	return ResourceWorkItemQueryPermissionsRead(d, m)
+}
+
+func ResourceWorkItemQueryPermissionsRead(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+
+	sn, err := securityhelper.NewSecurityNamespace(clients.Ctx,
+		securityhelper.SecurityNamespaceIDValues.WorkItemQueryFolders,
+		clients.SecurityClient,
+		clients.IdentityClient)
+	if err != nil {
+		return err
+	}
+
+	aclToken, err := createWorkItemQueryToken(clients.Ctx, clients.WorkItemTrackingClient, d)
+	if err != nil {
+		return err
+	}
+
+	principalPermissions, err := securityhelper.GetPrincipalPermissions(d, sn, aclToken)
+	if err != nil {
+		return err
+	}
+
+	d.Set("permissions", principalPermissions.Permissions)
+	return nil
+}
+
+func ResourceWorkItemQueryPermissionsUpdate(d *schema.ResourceData, m interface{}) error {
+	return ResourceWorkItemQueryPermissionsCreate(d, m)
```

Comment:
Can we merge `xxxxUpdate` and `xxxxCreate`  into one function `xxxxxxCreateOrUpdate`

Reply:
Of course ... Makes sense, because there's so little code inside `ResourceWorkItemQueryPermissionsUpdate`